tuxemon/core/components/networking.py

```diff
@@ -0,0 +1,777 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Tuxemon.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Contributor(s):
+#
+# William Edwards <shadowapex@gmail.com>
+# Derek Clark <derekjohn.clark@gmail.com>
+#
+# core.components.networking
+#
+"""This module contains the Tuxemon server and client.
+"""
+from middleware import Multiplayer
+from core.components import player
+from core.components.event.actions.npc import Npc
+from core import prepare
+
+from neteria.server import NeteriaServer
+from neteria.client import NeteriaClient
+
+import netifaces
+import pprint
+import pygame
+
+# Create a logger for optional handling of debug messages.
+import logging
+logger = logging.getLogger(__name__)
+pp = pprint.PrettyPrinter(indent=4)
+
+class TuxemonServer():
+    """Server class for multiplayer games. Creates a netaria server and
+    synchronizes the local game with all client states.
+
+    :param game: instance of the local game.
+    
+    :type game: core.tools.Control object.
+
+    :rtype: None
+    :returns: None
+
+    """
+    
+    def __init__(self, game):
+        self.game = game
+        self.server = NeteriaServer(Multiplayer(self))
+        self.network_events = []
+        self.listening = False
+        self.interfaces = {}
+        self.ips = []
+        
+        for device in netifaces.interfaces():
+            interface = netifaces.ifaddresses(device)
+            try:
+                self.interfaces[device] = interface[netifaces.AF_INET][0]['addr']
+            except KeyError:
+                pass
+        
+        for interface in self.interfaces:
+                        ip = self.interfaces[interface]
+                        if ip == "127.0.0.1": continue
+                        else: self.ips.append(ip)
+        
+
+    def update(self):
+        """Updates the server state with information sent from the clients
+
+        :param None
+        
+        :type None 
+
+        :rtype: None
+        :returns: None
+
+        """        
+        # Loop through our network events and pass them to the current state.
+        controller_events = self.net_controller_loop()
+        if controller_events:
+            for controller_event in controller_events:
+                self.game.key_events.append(controller_event)
+                self.game.state.get_event(controller_event)
+
+
+    def net_controller_loop(self):
+        """Process all network events from controllers and pass them
+        down to current State. All network events are converted to keyboard
+        events for compatibility.
+
+        :param None:
+
+        :rtype: None
+        :returns: None
+
+        """
+        events = []
+        for event_data in self.network_events:
+            if event_data == "KEYDOWN:up":
+                self.keys[pg.K_UP] = 1
```

Comment:
Fixed the issue and created a ControllerServer class that will start if net_controller_enabled = 1 in tuxemon.cfg. It uses the default port. The TuxemonServer has now been hard coded to 40081. The controller will need to be configured to only join a server with port 40080 or another port decided on later.